packages/legend-query/src/stores/QueryBuilderDateTimeHelper.ts

```diff
@@ -0,0 +1,587 @@
+  const latestStartDateValue = getStartDateValue(pureDateAjustFunction);
+  const dateAdjustOption: CustomDateOption = {
+    label: '',
+    value: QUERY_BUILDER_DATE_OPTION.CUSTOM_DATE,
+    unit: latestDurationUnitValue,
+    durationRange: latestDurationRangeValue,
+    startDate: latestStartDateValue,
+    adjustment: latestAdjustmentValue,
+  };
+
+  const matchedPreservedCustomAdjustDates = reservedCustomDates.filter(
+    (t) =>
+      guaranteeNonNullable(t.adjustment) +
+        guaranteeNonNullable(t.durationRange) +
+        guaranteeNonNullable(t.startDate) +
+        guaranteeNonNullable(t.unit) ===
+      guaranteeNonNullable(dateAdjustOption.adjustment) +
+        dateAdjustOption.durationRange +
+        dateAdjustOption.startDate +
+        dateAdjustOption.unit,
+  );
+  if (matchedPreservedCustomAdjustDates.length > 0) {
+    dateAdjustOption.label = guaranteeNonNullable(
+      matchedPreservedCustomAdjustDates[0]?.label,
+    );
+    dateAdjustOption.value = guaranteeNonNullable(
+      matchedPreservedCustomAdjustDates[0]?.value,
+    );
+    return dateAdjustOption;
+  }
+  dateAdjustOption.label = [
+    latestDurationRangeValue,
+    latestDurationUnitValue,
+    latestAdjustmentValue,
+    latestStartDateValue,
+  ].join(' ');
+  return dateAdjustOption;
+};
+
+export const generateCustomAjustDateOption = (
+  valueSpecification: SimpleFunctionExpression | PrimitiveInstanceValue,
+): CustomDateOption => {
+  if (
+    valueSpecification instanceof SimpleFunctionExpression &&
+    matchFunctionName(
+      valueSpecification.functionName,
+      SUPPORTED_FUNCTIONS.ADJUST,
+    )
+  ) {
+    return generateDateOptionForAdjustFunction(valueSpecification);
+  }
+  return {
+    label: '',
+    value: '',
+    durationRange: '',
+    unit: '',
+    adjustment: '',
+    startDate: '',
+  };
+};
+
+export const generateDateOptionForStartDayOfDateOption = (
+  unitValue: string,
+): CustomDateOption => {
+  switch (unitValue) {
+    case QUERY_BUILDER_DATE_DURATION_UNIT.WEEK:
+      return {
+        label: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_WEEK,
+        value: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_WEEK,
+      };
+    case QUERY_BUILDER_DATE_DURATION_UNIT.MONTH:
+      return {
+        label: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_MONTH,
+        value: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_MONTH,
+      };
+    case QUERY_BUILDER_DATE_DURATION_UNIT.QUARTER:
+      return {
+        label: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_QUARTER,
+        value: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_QUARTER,
+      };
+    case QUERY_BUILDER_DATE_DURATION_UNIT.YEAR:
+      return {
+        label: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_YEAR,
+        value: QUERY_BUILDER_CUSTOM_DATE_START_DATE_OPTION.START_OF_YEAR,
+      };
+    default:
+      return { label: '', value: '' };
```

Comment:
Does it hurt if instead of returning something like this, we return `undefined`?